File: node/src/query/planner.js
import { probeFileBloom } from '../bloom/bloom.js';

/**
 * Determines which files need to be scanned for a query.
 * Applies time-range filtering, then bloom filter pruning.
 */
export class QueryPlanner {
  constructor(coldStorage, dbCache) {
    this.coldStorage = coldStorage;
    this.dbCache = dbCache || null;
  }

  /**
   * Plan which files to query.
   * @param {object} manifest - tenant manifest with .files array
   * @param {object} query - { startTs, endTs, filters }
   * @returns {{ files: object[], stats: object }}
   */
  async plan(tenantId, manifest, query) {
    const { startTs, endTs, filters = {} } = query;
    const totalFiles = manifest.files.length;

    // 1. Time-range filter
    const timeFiltered = manifest.files.filter(
      (f) => f.endTs >= startTs && f.startTs <= endTs
    );

    // 2. Extract equality filters for bloom probing (skip wildcards/ranges)
    const bloomFilters = {};
    for (const [key, value] of Object.entries(filters)) {
      if (typeof value === 'string' && !value.includes('*') && !value.includes('?')) {
        bloomFilters[key] = value;
      } else if (Array.isArray(value) && value.every((v) => typeof v === 'string')) {
        bloomFilters[key] = value;
      }
      // skip range objects and wildcard patterns
    }

    // 3. Bloom filter pruning
    let bloomFiltered = timeFiltered;
    const hasBloomFilters = Object.keys(bloomFilters).length > 0;

    if (hasBloomFilters) {
      bloomFiltered = [];
      for (const file of timeFiltered) {
        const bloomData = await this._getBloomData(tenantId, file.segment);
        if (!bloomData || probeFileBloom(bloomData, bloomFilters)) {
          bloomFiltered.push(file);
        }
      }
    }

    return {
      files: bloomFiltered,
      stats: {
        totalFiles,
        filesAfterTimeFilter: timeFiltered.length,
        filesAfterBloom: bloomFiltered.length,
      },
    };
  }

  async _getBloomData(tenantId, segment) {
    // Try DuckDB cache first
    if (this.dbCache) {
      try {
        const data = await this.dbCache.getBloomData(tenantId, segment);
        if (data) return data;
      } catch {
        // fall through to cold storage
      }
    }

    // Fall back to cold storage
    try {
      return await this.coldStorage.getBloom(tenantId, segment);
    } catch {
      // No bloom available — can't prune
      return null;
    }
  }
}

File: go/internal/query/planner.go
package query

import (
	"fmt"
	"strings"

	"ducker/internal/bloom"
	"ducker/internal/model"
)

type BloomSource interface {
	GetBloomData(tenantID, segment string) (bloom.FileBloom, bool, error)
}

type ColdBloomSource interface {
	GetBloom(tenantID, segment string) (map[string]any, error)
}

type Planner struct {
	coldStorage ColdBloomSource
	dbCache     BloomSource
}

func NewPlanner(cold ColdBloomSource, cache BloomSource) *Planner {
	return &Planner{coldStorage: cold, dbCache: cache}
}

func (p *Planner) Plan(tenantID string, manifest model.Manifest, startTS, endTS int64, filters map[string]any) (model.QueryPlan, error) {
	totalFiles := len(manifest.Files)
	timeFiltered := make([]model.SegmentFile, 0, totalFiles)
	for _, f := range manifest.Files {
		if f.EndTS >= startTS && f.StartTS <= endTS {
			timeFiltered = append(timeFiltered, f)
		}
	}

	bloomFilters := extractBloomFilters(filters)
	hasBloomFilters := len(bloomFilters) > 0
	bloomFiltered := timeFiltered

	if hasBloomFilters {
		bloomFiltered = make([]model.SegmentFile, 0, len(timeFiltered))
		for _, file := range timeFiltered {
			data, ok, err := p.getBloomData(tenantID, file.Segment)
			if err != nil || !ok || bloom.ProbeFileBloom(data, bloomFilters) {
				bloomFiltered = append(bloomFiltered, file)
			}
		}
	}

	return model.QueryPlan{
		Files: bloomFiltered,
		Stats: model.PlanStats{
			TotalFiles:           totalFiles,
			FilesAfterTimeFilter: len(timeFiltered),
			FilesAfterBloom:      len(bloomFiltered),
		},
	}, nil
}

func extractBloomFilters(filters map[string]any) map[string]any {
	bloomFilters := map[string]any{}
	for k, v := range filters {
		s, ok := v.(string)
		if ok {
			if !strings.Contains(s, "*") && !strings.Contains(s, "?") {
				bloomFilters[k] = s
			}
			continue
		}
		if arr, ok := v.([]any); ok {
			allStrings := true
			vals := make([]string, 0, len(arr))
			for _, item := range arr {
				str, ok := item.(string)
				if !ok {
					allStrings = false
					break
				}
				vals = append(vals, str)
			}
			if allStrings {
				bloomFilters[k] = vals
			}
		}
	}
	return bloomFilters
}

func (p *Planner) getBloomData(tenantID, segment string) (bloom.FileBloom, bool, error) {
	if p.dbCache != nil {
		if data, ok, err := p.dbCache.GetBloomData(tenantID, segment); err == nil && ok {
			return data, true, nil
		}
	}
	raw, err := p.coldStorage.GetBloom(tenantID, segment)
	if err != nil {
		return bloom.FileBloom{}, false, nil
	}
	parsed, err := bloom.ParseFileBloom(raw)
	if err != nil {
		return bloom.FileBloom{}, false, fmt.Errorf("parse bloom: %w", err)
	}
	return parsed, true, nil
}
